crowbar_framework/app/controllers/api/upgrade_controller.rb

```diff
@@ -152,7 +152,12 @@ def nodes
         # When controller nodes have been upgraded previously,
         # whole 'nodes' step was not actually finished, just a substep.
         # It makes sense at this time to upgrade the rest with 'all'.
-        unless substep == :controller_nodes && status == :finished
+        unless (substep == :controller_nodes && status == :finished) ||
+            # Other case when we don't want to start the step again is
+            # when some compute node was already upgraded. Such case also leaves
+            # the 'nodes' step as running, but user might want to upgrade all
+            # remaining compute nodes by using 'all' argument.
```

Comment:
I do not like it as well, but to me this seems slightly better than big comment on the top.